{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "qduj991B0yQ2",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4c8a06ae-59cd-4289-e6f9-c66a7646a08a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rHit:1 https://cli.github.com/packages stable InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com] [Waiting for headers] [Connected to cloud\r                                                                               \rHit:2 http://security.ubuntu.com/ubuntu jammy-security InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com] [Waiting for headers] [Waiting for header\r                                                                               \rHit:3 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com (91.189.91.82)] [Waiting for headers] [Wai\r                                                                               \rHit:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com (91.189.91.82)] [Waiting for headers] [Con\r                                                                               \rHit:5 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:6 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Hit:7 http://archive.ubuntu.com/ubuntu jammy-updates InRelease\n",
            "Hit:8 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n",
            "Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "bcftools is already the newest version (1.13-1).\n",
            "samtools is already the newest version (1.13-4).\n",
            "tabix is already the newest version (1.13+ds-2build1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 46 not upgraded.\n",
            "Requirement already satisfied: pyfaidx in /usr/local/lib/python3.12/dist-packages (0.9.0.3)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.12/dist-packages (from pyfaidx) (25.0)\n"
          ]
        }
      ],
      "source": [
        "!apt-get update -y\n",
        "!apt-get install -y tabix bcftools samtools\n",
        "!pip install pyfaidx\n",
        "\n",
        "import gzip\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import os\n",
        "from pyfaidx import Fasta\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Skipped the 'subset to loci with matched transcript_ids in GRCh38.p14 GTF file' part from original paper\n",
        "\n",
        "CLINVAR_VCF_URL=\"https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/archive_2.0/2024/clinvar_20240301.vcf.gz\"\n",
        "#GTF_URL=\"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/latest_release/gencode.v49.annotation.gtf.gz\"\n",
        "\n",
        "DATA_DIR=\"/content/data\"\n",
        "!mkdir -p $DATA_DIR\n",
        "\n",
        "!wget -O $DATA_DIR/clinvar.vcf.gz \"$CLINVAR_VCF_URL\"\n",
        "#!wget -O $DATA_DIR/grch38_gtf.gtf.gz \"$GTF_URL\"\n",
        "\n",
        "!ls -lh $DATA_DIR\n",
        "\n",
        "!gunzip -c $DATA_DIR/clinvar.vcf.gz > $DATA_DIR/clinvar.vcf\n",
        "#!gunzip -c $DATA_DIR/grch38_gtf.gtf.gz > $DATA_DIR/grch38_gtf.gtf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qpDY-7XZLPh0",
        "outputId": "805d09ae-36a8-4ffa-d9c7-2b54fc5ad261"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-11-10 16:48:18--  https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/archive_2.0/2024/clinvar_20240301.vcf.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.13, 130.14.250.31, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 100783681 (96M) [application/x-gzip]\n",
            "Saving to: ‘/content/data/clinvar.vcf.gz’\n",
            "\n",
            "/content/data/clinv 100%[===================>]  96.11M  77.9MB/s    in 1.2s    \n",
            "\n",
            "2025-11-10 16:48:20 (77.9 MB/s) - ‘/content/data/clinvar.vcf.gz’ saved [100783681/100783681]\n",
            "\n",
            "total 1.2G\n",
            "-rw-r--r-- 1 root root 1.1G Nov 10 16:47 clinvar.vcf\n",
            "-rw-r--r-- 1 root root  97M Mar  1  2024 clinvar.vcf.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bcftools view $DATA_DIR/clinvar.vcf | head -n 60"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ciV1L2Hw5gik",
        "outputId": "fc8494d4-f183-482a-a81b-e49cbb0d99e3"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
            "##fileformat=VCFv4.1\n",
            "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
            "##fileDate=2024-03-01\n",
            "##source=ClinVar\n",
            "##reference=GRCh38\n",
            "##ID=<Description=\"ClinVar Variation ID\">\n",
            "##INFO=<ID=AF_ESP,Number=1,Type=Float,Description=\"allele frequencies from GO-ESP\">\n",
            "##INFO=<ID=AF_EXAC,Number=1,Type=Float,Description=\"allele frequencies from ExAC\">\n",
            "##INFO=<ID=AF_TGP,Number=1,Type=Float,Description=\"allele frequencies from TGP\">\n",
            "##INFO=<ID=ALLELEID,Number=1,Type=Integer,Description=\"the ClinVar Allele ID\">\n",
            "##INFO=<ID=CLNDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\n",
            "##INFO=<ID=CLNDNINCL,Number=.,Type=String,Description=\"For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">\n",
            "##INFO=<ID=CLNDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for germline classifications, e.g. OMIM:NNNNNN\">\n",
            "##INFO=<ID=CLNDISDBINCL,Number=.,Type=String,Description=\"For included Variant: Tag-value pairs of disease database name and identifier for germline classifications, e.g. OMIM:NNNNNN\">\n",
            "##INFO=<ID=CLNHGVS,Number=.,Type=String,Description=\"Top-level (primary assembly, alt, or patch) HGVS expression.\">\n",
            "##INFO=<ID=CLNREVSTAT,Number=.,Type=String,Description=\"ClinVar review status of germline classification for the Variation ID\">\n",
            "##INFO=<ID=CLNSIG,Number=.,Type=String,Description=\"Aggregate germline classification for this single variant; multiple values are separated by a vertical bar\">\n",
            "##INFO=<ID=CLNSIGCONF,Number=.,Type=String,Description=\"Conflicting germline classification for this single variant; multiple values are separated by a vertical bar\">\n",
            "##INFO=<ID=CLNSIGINCL,Number=.,Type=String,Description=\"Germline classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
            "##INFO=<ID=CLNVC,Number=1,Type=String,Description=\"Variant type\">\n",
            "##INFO=<ID=CLNVCSO,Number=1,Type=String,Description=\"Sequence Ontology id for variant type\">\n",
            "##INFO=<ID=CLNVI,Number=.,Type=String,Description=\"the variant's clinical sources reported as tag-value pairs of database and variant identifier\">\n",
            "##INFO=<ID=DBVARID,Number=.,Type=String,Description=\"nsv accessions from dbVar for the variant\">\n",
            "##INFO=<ID=GENEINFO,Number=1,Type=String,Description=\"Gene(s) for the variant reported as gene symbol:gene id. The gene symbol and id are delimited by a colon (:) and each pair is delimited by a vertical bar (|)\">\n",
            "##INFO=<ID=MC,Number=.,Type=String,Description=\"comma separated list of molecular consequence in the form of Sequence Ontology ID|molecular_consequence\">\n",
            "##INFO=<ID=ONCDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDB\">\n",
            "##INFO=<ID=ONCDNINCL,Number=.,Type=String,Description=\"For included variant: ClinVar's preferred disease name for the concept specified by disease identifiers in ONCDISDBINCL\">\n",
            "##INFO=<ID=ONCDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for oncogenicity classifications, e.g. MedGen:NNNNNN\">\n",
            "##INFO=<ID=ONCDISDBINCL,Number=.,Type=String,Description=\"For included variant: Tag-value pairs of disease database name and identifier for oncogenicity classifications, e.g. OMIM:NNNNNN\">\n",
            "##INFO=<ID=ONC,Number=.,Type=String,Description=\"Aggregate oncogenicity classification for this single variant; multiple values are separated by a vertical bar\">\n",
            "##INFO=<ID=ONCINCL,Number=.,Type=String,Description=\"Oncogenicity classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
            "##INFO=<ID=ONCREVSTAT,Number=.,Type=String,Description=\"ClinVar review status of oncogenicity classification for the Variation ID\">\n",
            "##INFO=<ID=ONCCONF,Number=.,Type=String,Description=\"Conflicting oncogenicity classification for this single variant; multiple values are separated by a vertical bar\">\n",
            "##INFO=<ID=ORIGIN,Number=.,Type=String,Description=\"Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other\">\n",
            "##INFO=<ID=RS,Number=.,Type=String,Description=\"dbSNP ID (i.e. rs number)\">\n",
            "##INFO=<ID=SCIDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in SCIDISDB\">\n",
            "##INFO=<ID=SCIDNINCL,Number=.,Type=String,Description=\"For included variant: ClinVar's preferred disease name for the concept specified by disease identifiers in SCIDISDBINCL\">\n",
            "##INFO=<ID=SCIDISDB,Number=.,Type=String,Description=\"Tag-value pairs of disease database name and identifier submitted for somatic clinial impact classifications, e.g. MedGen:NNNNNN\">\n",
            "##INFO=<ID=SCIDISDBINCL,Number=.,Type=String,Description=\"For included variant: Tag-value pairs of disease database name and identifier for somatic clinical impact classifications, e.g. OMIM:NNNNNN\">\n",
            "##INFO=<ID=SCIREVSTAT,Number=.,Type=String,Description=\"ClinVar review status of somatic clinical impact for the Variation ID\">\n",
            "##INFO=<ID=SCI,Number=.,Type=String,Description=\"Aggregate somatic clinical impact for this single variant; multiple values are separated by a vertical bar\">\n",
            "##INFO=<ID=SCIINCL,Number=.,Type=String,Description=\"Somatic clinical impact classification for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:classification; multiple values are separated by a vertical bar\">\n",
            "##bcftools_viewVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_viewCommand=view /content/data/clinvar.vcf; Date=Mon Nov 10 16:48:26 2025\n",
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n",
            "1\t69134\t2205837\tA\tG\t.\t.\tALLELEID=2193183;CLNDISDB=MeSH:D030342,MedGen:C0950123;CLNDN=Inborn_genetic_diseases;CLNHGVS=NC_000001.11:g.69134A>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1\n",
            "1\t69581\t2252161\tC\tG\t.\t.\tALLELEID=2238986;CLNDISDB=MeSH:D030342,MedGen:C0950123;CLNDN=Inborn_genetic_diseases;CLNHGVS=NC_000001.11:g.69581C>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1\n",
            "1\t69682\t2396347\tG\tA\t.\t.\tALLELEID=2386655;CLNDISDB=MeSH:D030342,MedGen:C0950123;CLNDN=Inborn_genetic_diseases;CLNHGVS=NC_000001.11:g.69682G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1\n",
            "1\t69769\t2288999\tT\tC\t.\t.\tALLELEID=2278803;CLNDISDB=MeSH:D030342,MedGen:C0950123;CLNDN=Inborn_genetic_diseases;CLNHGVS=NC_000001.11:g.69769T>C;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1\n",
            "1\t69995\t2351346\tG\tC\t.\t.\tALLELEID=2333177;CLNDISDB=MeSH:D030342,MedGen:C0950123;CLNDN=Inborn_genetic_diseases;CLNHGVS=NC_000001.11:g.69995G>C;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=OR4F5:79501;MC=SO:0001583|missense_variant;ORIGIN=1\n",
            "1\t925946\t1924157\tC\tG\t.\t.\tALLELEID=1983057;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925946C>G;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1\n",
            "1\t925952\t1019397\tG\tA\t.\t.\tALLELEID=1003021;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925952G>A;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1640863258\n",
            "1\t925956\t1543320\tC\tT\t.\t.\tALLELEID=1632777;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925956C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001819|synonymous_variant;ORIGIN=1;RS=1342334044\n",
            "1\t925961\t2069387\tA\tT\t.\t.\tALLELEID=2129477;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925961A>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1\n",
            "1\t925969\t1648427\tC\tT\t.\t.\tAF_EXAC=0.00033;ALLELEID=1600580;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925969C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1;RS=200686669\n",
            "1\t925976\t1362713\tT\tC\t.\t.\tALLELEID=1396033;CLNDISDB=MeSH:D030342,MedGen:C0950123|MedGen:C3661900;CLNDN=Inborn_genetic_diseases|not_provided;CLNHGVS=NC_000001.11:g.925976T>C;CLNREVSTAT=criteria_provided,_multiple_submitters,_no_conflicts;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1030245330\n",
            "1\t925980\t1936488\tC\tT\t.\t.\tALLELEID=1986319;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925980C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001819|synonymous_variant;ORIGIN=1\n",
            "1\t925986\t1568423\tC\tT\t.\t.\tALLELEID=1570515;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.925986C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001819|synonymous_variant;ORIGIN=1;RS=1640864832\n",
            "1\t926003\t1365270\tC\tT\t.\t.\tALLELEID=1502313;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.926003C>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Uncertain_significance;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001583|missense_variant;ORIGIN=1;RS=1329301928\n",
            "1\t926010\t1632958\tG\tT\t.\t.\tALLELEID=1545352;CLNDISDB=MedGen:C3661900;CLNDN=not_provided;CLNHGVS=NC_000001.11:g.926010G>T;CLNREVSTAT=criteria_provided,_single_submitter;CLNSIG=Likely_benign;CLNVC=single_nucleotide_variant;CLNVCSO=SO:0001483;GENEINFO=SAMD11:148398;MC=SO:0001819|synonymous_variant;ORIGIN=1;RS=889842126\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bcftools query -f '%CHROM\\t%POS\\t%ID\\t%REF\\t%ALT\\t%INFO/ALLELEID\\t%INFO/GENEINFO\\t%INFO/CLNSIG\\t%INFO/CLNREVSTAT\\t%INFO/RS\\t%INFO/CLNHGVS\\n' \\\n",
        "  $DATA_DIR/clinvar.vcf > clinvar_raw.tsv 2>/dev/null\n",
        "!head -n 10 clinvar_raw.tsv"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Mfr7wN61jBVY",
        "outputId": "277ef6d8-9d10-4e8d-f926-58346d92bd32"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1\t69134\t2205837\tA\tG\t2193183\tOR4F5:79501\tLikely_benign\tcriteria_provided,_single_submitter\t.\tNC_000001.11:g.69134A>G\n",
            "1\t69581\t2252161\tC\tG\t2238986\tOR4F5:79501\tUncertain_significance\tcriteria_provided,_single_submitter\t.\tNC_000001.11:g.69581C>G\n",
            "1\t69682\t2396347\tG\tA\t2386655\tOR4F5:79501\tUncertain_significance\tcriteria_provided,_single_submitter\t.\tNC_000001.11:g.69682G>A\n",
            "1\t69769\t2288999\tT\tC\t2278803\tOR4F5:79501\tUncertain_significance\tcriteria_provided,_single_submitter\t.\tNC_000001.11:g.69769T>C\n",
            "1\t69995\t2351346\tG\tC\t2333177\tOR4F5:79501\tUncertain_significance\tcriteria_provided,_single_submitter\t.\tNC_000001.11:g.69995G>C\n",
            "1\t925946\t1924157\tC\tG\t1983057\tSAMD11:148398\tUncertain_significance\tcriteria_provided,_single_submitter\t.\tNC_000001.11:g.925946C>G\n",
            "1\t925952\t1019397\tG\tA\t1003021\tSAMD11:148398\tUncertain_significance\tcriteria_provided,_single_submitter\t1640863258\tNC_000001.11:g.925952G>A\n",
            "1\t925956\t1543320\tC\tT\t1632777\tSAMD11:148398\tLikely_benign\tcriteria_provided,_single_submitter\t1342334044\tNC_000001.11:g.925956C>T\n",
            "1\t925961\t2069387\tA\tT\t2129477\tSAMD11:148398\tUncertain_significance\tcriteria_provided,_single_submitter\t.\tNC_000001.11:g.925961A>T\n",
            "1\t925969\t1648427\tC\tT\t1600580\tSAMD11:148398\tLikely_benign\tcriteria_provided,_single_submitter\t200686669\tNC_000001.11:g.925969C>T\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import re\n",
        "\n",
        "cols = [\"CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"ALLELEID\",\"GENEINFO\",\"CLNSIG\",\"CLNREVSTAT\",\"RS\",\"CLNHGVS\"]\n",
        "df = pd.read_csv(\"clinvar_raw.tsv\", sep=\"\\t\", names=cols, comment=\"#\", dtype=str)\n",
        "\n",
        "# Filtering\n",
        "\n",
        "# Keep nuclear chromosomes only\n",
        "chr = df['CHROM'].str.match(r'^(chr)?([1-9]|1[0-9]|2[0-2]|X|Y)$')\n",
        "df = df[chr].copy()\n",
        "# Keep SNVs only\n",
        "df = df[df['REF'].str.len() == 1]\n",
        "df = df[df['ALT'].str.len() == 1].copy()\n",
        "# Review status (at least two stars)\n",
        "rev_re = re.compile(r'practice_guideline|reviewed_by_expert_panel|criteria_provided,_multiple_submitters')\n",
        "df = df[df['CLNREVSTAT'].astype(str).str.contains(rev_re)]\n",
        "# Exclude VUS and collapse CLNSIG\n",
        "df = df[~df['CLNSIG'].astype(str).str.contains('Uncertain_significance', na=False)]\n",
        "P = df['CLNSIG'].str.contains('Pathogenic|Likely_pathogenic', na=False)\n",
        "B = df['CLNSIG'].str.contains('Benign|Likely_benign', na=False)\n",
        "df = df[(P ^ B)]\n",
        "df.loc[P, 'LABEL'] = 'Pathogenic/Likely_pathogenic'\n",
        "df.loc[B, 'LABEL'] = 'Benign/Likely_benign'\n",
        "\n",
        "df.to_csv(\"clinvar_filtered.tsv\", sep=\"\\t\", index=False)\n",
        "print(\"clinvar_filtered.tsv! rows:\", len(df))\n",
        "\n",
        "!head -n 10 clinvar_filtered.tsv"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TvNQWSi5j_9L",
        "outputId": "d2631a01-9a61-48a0-fce2-8b34888bc6c7"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "clinvar_filtered.tsv! rows: 192949\n",
            "CHROM\tPOS\tID\tREF\tALT\tALLELEID\tGENEINFO\tCLNSIG\tCLNREVSTAT\tRS\tCLNHGVS\tLABEL\n",
            "1\t935779\t1167937\tG\tA\t1153708\tSAMD11:148398\tBenign/Likely_benign\tcriteria_provided,_multiple_submitters,_no_conflicts\t143282473\tNC_000001.11:g.935779G>A\tBenign/Likely_benign\n",
            "1\t944036\t1154628\tG\tA\t1131808\tSAMD11:148398\tLikely_benign\tcriteria_provided,_multiple_submitters,_no_conflicts\t202045502\tNC_000001.11:g.944036G>A\tBenign/Likely_benign\n",
            "1\t963233\t775485\tC\tT\t696923\tKLHL17:339451\tBenign\tcriteria_provided,_multiple_submitters,_no_conflicts\t61746776\tNC_000001.11:g.963233C>T\tBenign/Likely_benign\n",
            "1\t970514\t710495\tC\tT\t730039\tPLEKHN1:84069\tBenign/Likely_benign\tcriteria_provided,_multiple_submitters,_no_conflicts\t41285810\tNC_000001.11:g.970514C>T\tBenign/Likely_benign\n",
            "1\t1013490\t1185393\tC\tG\t1175250\tISG15:9636\tBenign\tcriteria_provided,_multiple_submitters,_no_conflicts\t4615788\tNC_000001.11:g.1013490C>G\tBenign/Likely_benign\n",
            "1\t1013541\t1185394\tT\tC\t1175251\tISG15:9636\tBenign\tcriteria_provided,_multiple_submitters,_no_conflicts\t15842\tNC_000001.11:g.1013541T>C\tBenign/Likely_benign\n",
            "1\t1014052\t1577559\tG\tA\t1553595\tISG15:9636\tLikely_benign\tcriteria_provided,_multiple_submitters,_no_conflicts\t939961004\tNC_000001.11:g.1014052G>A\tBenign/Likely_benign\n",
            "1\t1014217\t475278\tC\tT\t446987\tISG15:9636\tBenign/Likely_benign\tcriteria_provided,_multiple_submitters,_no_conflicts\t61766284\tNC_000001.11:g.1014217C>T\tBenign/Likely_benign\n",
            "1\t1014228\t402986\tG\tA\t389314\tISG15:9636\tBenign\tcriteria_provided,_multiple_submitters,_no_conflicts\t1921\tNC_000001.11:g.1014228G>A\tBenign/Likely_benign\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# The original filtered ClinVar set is too large (192949 rows).\n",
        "# Randomly sample from it to get a subset of a reasonable size\n",
        "sample_df = df.sample(frac=0.05, random_state=42)\n",
        "sample_df.to_csv(\"clinvar_subset.tsv\", sep=\"\\t\", index=False)\n",
        "# Export the dataset and later feed it into Evo 2"
      ],
      "metadata": {
        "id": "EU_obPPA61xy"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Extract TP53 variants to get a TP53 test set\n",
        "df_tp53 = df[df[\"GENEINFO\"].astype(str).str.contains(\"TP53\", case=True, na=False)].reset_index(drop=True)\n",
        "print(\"Total TP53 variants:\", len(df_tp53))\n",
        "df_tp53.to_csv(\"clinvar_tp53_test.tsv\", sep=\"\\t\", index=False)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zoWJD3EFqMgE",
        "outputId": "6a4658ea-4149-4670-da29-bcb8c4b74599"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total TP53 variants: 610\n"
          ]
        }
      ]
    }
  ]
}